[WIP] Add new selector for Chargeback Rate editing screen to select the base model

hstastna opened the pull request:
**Partially fixes:**
https://bugzilla.redhat.com/show_bug.cgi?id=1581817
**Needs to merge with:**
https://github.com/ManageIQ/manageiq/pull/17507
https://github.com/ManageIQ/manageiq-schema/pull/304 (new alternative of https://github.com/ManageIQ/manageiq-schema/pull/209)

---

**What:**
When choosing _Chargeback for Images/Projects/Vms_ under _Configure Report Columns_ section while adding a new/editing an existing Report under _Cloud Intel > Reports > Reports_, some of the fields are not available. This is ok but there was a misunderstanding that those fields were missing there. This is why I add note/info about which fields are not supported to the Report editing screen (in another PR: https://github.com/ManageIQ/manageiq-ui-classic/pull/4081), and also new selector for Chargeback Rate editing screen (under _Cloud Intel > Chargeback > Rates_) is added to choose what the Rate is based on (_Images/Projects/Vms_) (this PR). We call it the base model (see https://github.com/ManageIQ/manageiq-schema/pull/209).

---

**Before:**
Adding/editing a Chargeback Rate:
![edit_rate_before](https://user-images.githubusercontent.com/13417815/40666135-500ed858-635f-11e8-9285-78adf59e6e15.png)
Displaying details page of a Chargeback Rate:
![rate_show](https://user-images.githubusercontent.com/13417815/41099016-872b10ea-6a5d-11e8-9aa5-f1c7e619a2d8.png)

---

**After:**
Adding a new Chargeback Rate (the _Rate Details_ table is displayed after choosing one of the "base models" by the new drop down because the rate details depend on it):
![add_new_rate_after](https://user-images.githubusercontent.com/13417815/41103181-c8b2e5c8-6a68-11e8-87bb-fdf5fed27474.png)
The new selector for configuring Rate columns (for choosing the base model), in editing screen:
![report_base_model-select](https://user-images.githubusercontent.com/13417815/41103191-cf2e49ba-6a68-11e8-9245-afe7a9fff787.png)
Adding/editing an existing Chargeback Rate with base model:
(After changing the model, the Rate Details table is cleared to its default values, similarly as for Reports, and the columns of the table are updated according to the chosen base model: only appropriate columns are displayed for each base model)
![new_rate_base_model](https://user-images.githubusercontent.com/13417815/41104459-ae143cdc-6a6b-11e8-8066-b4e7e4f5f622.png)
Displaying details page of a new Chargeback Rate, with the base model:
![new_rate2_show](https://user-images.githubusercontent.com/13417815/41104589-0489b808-6a6c-11e8-8f2d-3cb42cfc2061.png)
Displaying details page of a Chargeback Rate, **without the base model** (all the existing rates):
(The rate is based on 'Any' as its details' table has all the available/any columns)
![old_rate_show](https://user-images.githubusercontent.com/13417815/41051078-65ff2abc-69b5-11e8-9819-eaf604827e9f.png)
Editing the Chargeback Rate without the base model:
('Any' option is also available - user can see all the original details' values and edit them)
![old_rate-edit](https://user-images.githubusercontent.com/13417815/41104808-8cd03a7a-6a6c-11e8-8504-b73d752c8a4e.png)
Changing the base model of the Chargeback Rate without the base model:
(While changing the base model from "Any" to any other, user will not lose his original details' values, the table will not be reset to its default values as usually, only the appropriate columns display - with already existing values of the rate, already saved in the DB)
![old_rate_change_base](https://user-images.githubusercontent.com/13417815/41051612-e5f76f62-69b6-11e8-841d-2c3d1a8b46eb.png)
Saving the Chargeback Rate without the base model:
![model_required](https://user-images.githubusercontent.com/13417815/41051452-77b61904-69b6-11e8-946c-15f41c536c63.png)

**Note:**
Copying of the Rate with/without base model also works but it will not be possible to save the Rate based on 'Any'.

lpichler commented:
@miq-bot assign @mzazrivec

skateman commented on app/views/chargeback/_cb_rate_edit.html.haml:
```diff
@@ -36,4 +36,10 @@
+    %br
+    = _('* Caution: Allocated CPU Count/Cores, Used CPU, Used Disk I/O and Allocated Memory are not supported for Chargeback for Projects. ')
+    %br
+    = _('* Caution: Allocated/Used CPU Cores are not supported for Chargeback for Vms. ')
```
Could we just rephrase these into a single sentence on a single line?

lpichler replied:
I don't understand how. These are three different sentences how would you like to put it into one ?

skateman replied:
keyword: rephrase :wink:

hstastna replied:
There are three things: Chargeback for Images, Projects, Vms. We need to inform which fields are not supported for each of them. How would you put so much info into the one line to keep readability? Anyway, I see later comments and it looks like there will be completely different solution to what I was told to do here.

skateman replied:
Well, for me it looks weird to have a lot of warning signs...something like this in real life:
![](http://www.signs4safety.com/SIGNS4SAFETYCATALOG/New%20Folder/automotivesigns5.jpg)

hstastna replied:
Well, four lines of "Caution" is also too much for me but here it is not possible to put all the info in one line and to keep readability, it's too much info to put it just into one line. Anyway, I was told to do so.. Aaand.. now I was told to do it completely differently so don't worry about this, I will udpate the PR soon! ;)

skateman commented on app/views/report/_form_columns.html.haml:
```diff
@@ -26,6 +26,13 @@
+        - report_base = @edit[:new][:model].titleize.split(' ').last.pluralize
+        %br
+        %strong
+          = _("* Caution: #{unavailable_fields} are not supported for Chargeback for #{report_base}.")
```
This is probably not okay from the translation side, you need to use the `%` operator.

skateman commented on app/views/report/_form_columns.html.haml:
```diff
@@ -26,6 +26,13 @@
+    - if @edit[:new][:model] && @edit[:new][:model].starts_with?('Chargeback')
+      - unavailable_fields = unavailable_fields_for_model
+      - if unavailable_fields
+        - report_base = @edit[:new][:model].titleize.split(' ').last.pluralize
```
Translation missing here, probably `ui_lookup` but not sure.

hstastna replied:
Here I don't display anything so I don't think so. I think I miss `ui_lookup` later.

skateman replied:
You're displaying `report_base` a few lines below.

hstastna replied:
Yes, this is what i meant by "later".

skateman commented on app/controllers/report_controller.rb:
```diff
@@ -274,6 +274,19 @@ def title
 
+  # Get string with unavailable fields for displaying note about them while adding/editing report
+  def unavailable_fields_for_model
+    case @edit[:new][:model]
```
Instead of testing for the `@edit[:new][:model]` I would make this method more *universal* by using an argument instead.

skateman commented on app/controllers/report_controller.rb:
```diff
@@ -274,6 +274,19 @@ def title
+      'Allocated CPU Count, Used CPU, Used Disk I/O and Used Network I/O'
+    when 'ChargebackContainerProject'
+      'Allocated CPU Count/Cores, Used CPU, Used Disk I/O and Allocated Memory'
+    end
```
Translations?

hstastna replied:
I wouldn't use `ui_lookup` here. Or did you mean anything else?

skateman replied:
`_`

lpichler commented:
after clarification with @martinpovolny we will do the different solution.  It will be about "Chargeback Rates for concrete type of report"

hstastna commented:
I've updated this PR: changed displaying info about unavailable fields while adding/editing a report using method for getting the unavailable fields from backend, removed unrelated commits, updated the description, screenshots. TODO: adding/editing a Chargeback Rate.

hstastna commented:
I've finished the changes for Reports and also Rates. Everything should work. Specs are not added as this PR will probably need some small changes (after some reviewing).

hstastna commented:
Changes for Reports moved to another PR.
